### PR TITLE
[Spark] Remove waiting for a fixed time for Delta Connect Server to be available in Delta Connect testing

### PR DESCRIPTION
**Pull request by longvu-db**
#### Which Delta project/connector is this regarding?

- [X] Spark
- [ ] Standalone
- [ ] Flink
- [ ] Kernel
- [ ] Other (fill in here)

## Description
For local E2E Delta Connect testing, we also designed an [util class](https://github.com/delta-io/delta/blob/01bf60743b77c47147843e9083129320490f1629/spark-connect/client/src/test/scala-spark-master/io/delta/connect/tables/RemoteSparkSession.scala#L62) to start a local server in a different process similar to [SparkConnect](https://github.com/apache/spark/blob/ba208b9ca99990fa329c36b28d0aa2a5f4d0a77e/connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/test/RemoteSparkSession.scala#L37).

We noticed that the server takes a random amount of seconds to start up, and back then we received the error `INVALID_HANDLE.SESSION_NOT_FOUND] The handle 746e6c86-9fa9-4b08-9572-388c20eaed47 is invalid. Session not found. SQLSTATE: HY000"`, so what we did is to add a 10s `Thread.sleep` before starting the client.

This is not robust, so we are removing the `Thread.sleep`. This should work because:
1. The SparkSession's builder here already uses the default [Configuration](https://github.com/apache/spark/blob/3edc9c23a723a92c5a951cea0436529de65c640a/connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala#L891) of the `SparkConnectClient` which includes a default retry policy.
2. Spark patches the error `INVALID_HANDLE.SESSION_NOT_FOUND` in this [PR](https://github.com/apache/spark/pull/46971) at some point, so we should be able to retry even if encountering this error.

## How was this patch tested?
Existing UTs.

Ran `build/sbt` tests locally at least 20 times.

## Does this PR introduce _any_ user-facing changes?

No.

### CODE REVIEW

File: spark-connect/client/src/test/scala-spark-master/io/delta/connect/tables/RemoteSparkSession.scala [hunk 1]
```diff
@@ -97,10 +98,14 @@ trait RemoteSparkSession extends BeforeAndAfterAll { self: Suite =>
...
-    Thread.sleep(10000)
-    spark = SparkSession.builder().remote(s"sc://localhost:$serverPort").build()
+
+    val sparkConnectClient = SparkConnectClient
+      .builder()
+      .userId("test")
+      .port(serverPort)
+      .retryPolicy(RetryPolicy.defaultPolicy())
+
+    spark = SparkSession.builder().client(sparkConnectClient.build()).create()
```
Comment (longvu-db) [inline spark-connect/client/src/test/scala-spark-master/io/delta/connect/tables/RemoteSparkSession.scala:108]: `build` is deprecated - https://github.com/apache/spark/blob/3edc9c23a723a92c5a951cea0436529de65c640a/connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala#L1015

File: spark-connect/client/src/test/scala-spark-master/io/delta/connect/tables/DeltaTableSuite.scala [hunk 2]
```diff
@@ -19,6 +19,8 @@ package io.delta.tables
 import java.io.File
 import java.text.SimpleDateFormat
 
+import org.apache.spark.sql.delta.tables.RemoteSparkSession
```
Comment (tomvanbussel) [inline spark-connect/client/src/test/scala-spark-master/io/delta/connect/tables/DeltaTableSuite.scala:22]: Unrelated change?

File: spark-connect/client/src/test/scala-spark-master/io/delta/connect/tables/RemoteSparkSession.scala [hunk 3]
```diff
@@ -97,10 +98,14 @@ trait RemoteSparkSession extends BeforeAndAfterAll { self: Suite =>
...
-    // TODO: Instead of sleeping for a fixed time, which is a bit brittle,
-    // we should repeatedly check when the server is ready.
-    Thread.sleep(10000)
-    spark = SparkSession.builder().remote(s"sc://localhost:$serverPort").build()
+
+    val sparkConnectClient = SparkConnectClient
+      .builder()
+      .userId("test")
+      .port(serverPort)
+      .retryPolicy(RetryPolicy.defaultPolicy())
```
Comment (tomvanbussel) [inline spark-connect/client/src/test/scala-spark-master/io/delta/connect/tables/RemoteSparkSession.scala:106]: I don't think this is necessary. This is already the default.

File: spark-connect/client/src/test/scala-spark-master/io/delta/connect/tables/RemoteSparkSession.scala (hunk 1 above)
Comment (tomvanbussel) [inline spark-connect/client/src/test/scala-spark-master/io/delta/connect/tables/RemoteSparkSession.scala:108]: Why can't we just replace `.build()` with `.create()`? Why do we need to provide our own `SparkConnectClient`?

**Review by tomvanbussel — COMMENTED** [review]
Not sure why these changes are necessary. The retry policy is already the default. Wouldn't just removing the sleep be enough?

File: spark-connect/client/src/test/scala-spark-master/io/delta/connect/tables/RemoteSparkSession.scala (hunk 3 above)
Comment (longvu-db) [reply spark-connect/client/src/test/scala-spark-master/io/delta/connect/tables/RemoteSparkSession.scala:106]: @tomvanbussel I put it here because I feel like the fact that is we are using the `defaultPolicy` might not be obvious and the viewers might not know, should I put a comment instead?

File: spark-connect/client/src/test/scala-spark-master/io/delta/connect/tables/DeltaTableSuite.scala (hunk 2 above)
Comment (longvu-db) [reply spark-connect/client/src/test/scala-spark-master/io/delta/connect/tables/DeltaTableSuite.scala:22]: We changed the package of `RemoteSparkSession`from `io.delta.tables` to ` org.apache.spark.sql.delta.tables` because `SparkConnectClient` is `private[sql]`.

https://github.com/apache/spark/blob/3edc9c23a723a92c5a951cea0436529de65c640a/sql/connect/common/src/main/scala/org/apache/spark/sql/connect/client/SparkConnectClient.scala#L40

So now `DeltaTableSuite` doesn't have the same package name as `RemoteSparkSession`, we have to include explicitly.

File: spark-connect/client/src/test/scala-spark-master/io/delta/connect/tables/RemoteSparkSession.scala (hunk 1 above)
Comment (longvu-db) [reply spark-connect/client/src/test/scala-spark-master/io/delta/connect/tables/RemoteSparkSession.scala:108]: @tomvanbussel 

https://github.com/apache/spark/blob/3edc9c23a723a92c5a951cea0436529de65c640a/connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala#L1028

If we only do like this,

`SparkSession.builder().remote(s"sc://localhost:$serverPort").create()`

The only config it takes into account is `remote`

We need to include `client` to have `RetryPolicy` from `SparkConnectClient`

https://github.com/apache/spark/blob/3edc9c23a723a92c5a951cea0436529de65c640a/sql/connect/common/src/main/scala/org/apache/spark/sql/connect/client/SparkConnectClient.scala#L722

`SparkSession`'s builder doesn't have `RetryPolicy`
https://github.com/apache/spark/blob/3edc9c23a723a92c5a951cea0436529de65c640a/connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala#L887

Similar to here
https://github.com/apache/spark/blob/3edc9c23a723a92c5a951cea0436529de65c640a/connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/test/RemoteSparkSession.scala#L181

File: spark-connect/client/src/test/scala-spark-master/io/delta/connect/tables/RemoteSparkSession.scala (hunk 3 above)
Comment (tomvanbussel) [reply spark-connect/client/src/test/scala-spark-master/io/delta/connect/tables/RemoteSparkSession.scala:106]: Unfortunately it requires accessing a package private interface. I would rather use the public interfaces, even if that means not being explicit about using the default retry policy.

File: spark-connect/client/src/test/scala-spark-master/io/delta/connect/tables/RemoteSparkSession.scala (hunk 1 above)
Comment (tomvanbussel) [reply spark-connect/client/src/test/scala-spark-master/io/delta/connect/tables/RemoteSparkSession.scala:108]: `remote` still uses the default retry policy, right?

Comment (longvu-db) [reply spark-connect/client/src/test/scala-spark-master/io/delta/connect/tables/RemoteSparkSession.scala:108]: @tomvanbussel You are right, by default the `SparkSession`'s `builder` here already uses the default `Configuration` of the `SparkConnectClient` (https://github.com/apache/spark/blob/3edc9c23a723a92c5a951cea0436529de65c640a/connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala#L891).

I thought that the old approach would not work because it was failing back then, with this error
`INVALID_HANDLE.SESSION_NOT_FOUND] The handle 746e6c86-9fa9-4b08-9572-388c20eaed47 is invalid. Session not found. SQLSTATE: HY000"`, but Spark patches this [error](https://github.com/apache/spark/pull/46971) a week later by still retrying even when encountering this exception, so our way back then should work even without the `Thread.sleep`

**Review by tomvanbussel — APPROVED** [review]
